### PR TITLE
Refactor complex LoginController logic into a login chain of responsibility

### PR DESCRIPTION
**Pull request by ChristophWurst**
This method has been on my list of code to refactor for a long time. As I now have to touch it for https://github.com/orgs/nextcloud/projects/29, I decided to break the code into smaller, maintainable and testable chunks. The code now follows a more strict single responsibility principle, while implementing the  [chain of responsibility pattern](https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern).

### CODE REVIEW

**Comment by ChristophWurst** [comment]
From what I can tell CI failures do not look related. But tell me if I'm wrong.

**Comment by ChristophWurst** [comment]
> Redis::connect(): connect() failed: Connection refused 

-> restarted

**Review by rullzer — APPROVED** [review]
This is just great. So much cleaner. And tests :rocket: 

I didn't test in debt but most stuff didn't blow up with my smoke testing.

File: tests/lib/Authentication/Login/TwoFactorCommandTest.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		$this->urlGenerator = $this->createMock(IURLGenerator::class);
+
+		$this->cmd = new TwoFactorCommand(
+			$this->twoFactorManager,
+			$this->urlGenerator
+		);
+	}
+
+	public function testNotTwoFactorAuthenticated() {
+		$this->fail('todo');
```
Comment (ChristophWurst) [inline tests/lib/Authentication/Login/TwoFactorCommandTest.php]: guess what … CI fails for a reason

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

File: tests/lib/Authentication/Login/TwoFactorCommandTest.php (hunk 1 above)
Comment (rullzer) [reply tests/lib/Authentication/Login/TwoFactorCommandTest.php]: NO WAY!

**Pull request merged**